python/lsst/daf/butler/tests/_testRepo.py

```diff
@@ -370,45 +453,33 @@ def addDataIdValue(butler: Butler, dimension: str, value: Union[str, int], **rel
     for dimension_obj in all_dimensions:
         dimension_value = data_id.get(dimension_obj.name)
         if dimension_value is None:
-            # No values given by caller for this dimension.  See if any exist
-            # in the registry that are consistent with the values of dimensions
-            # we do have:
-            match_data_id = {key: data_id[key] for key in data_id.keys() & dimension_obj.dimensions.names}
-            matches = list(
-                butler.registry.queryDimensionRecords(dimension_obj, dataId=match_data_id).limit(1)
-            )
-            if not matches:
-                # Nothing in the registry matches: invent a data ID value
-                # with the right type (actual value does not matter).
-                dimension_value = _makeRandomDataIdValue(dimension_obj)
-                data_id[dimension_obj.name] = dimension_value
-            else:
-                # A record does exist in the registry.  Use its data ID value,
-                # and skip the rest of the record-fabrication logic.
-                data_id[dimension_obj.name] = matches[0].dataId[dimension_obj.name]
+            data_id[dimension_obj.name], invented = _findOrInventDataIdValue(butler, data_id, dimension_obj)
+            if not invented:
+                # No need to make a new record; one already exists.
                 continue
+        if dimension_obj.name in related:
+            # Caller passed in a value of this dimension explicitly, but it
+            # isn't the primary dimension they asked to have a record created
+            # for.  That means they expect this record to already exist.
+            continue
+        if dimension_obj != fullDimension and dimension_obj in all_dimensions.required:
+            # We also don't want to automatically create new dimension records
+            # for required dimensions (except for the main dimension the caller
+            # asked for); those are also asserted by the caller to already
+            # exist.
+            continue
```

Comment:
This was the original behavior, but I thought (based on your answer to https://github.com/lsst/daf_butler/pull/680#discussion_r871800711) that this wouldn't work anymore. Given the incoming not-null constraints, what "non-obvious" dimensions will people need to specify that they didn't before? `physical_filter` for `exposure`?

Reply:
The original behavior was that users did not need to provide values for implied dependencies, and the foreign key columns with the values for those dimensions were set to `None` / `NULL`.

The new behavior is that users still do not need to provide values for implied dependencies, but this is now because those missing values are generated randomly and records for them are inserted into the registry, allowing the foreign key columns that point to them to be not-`NULL` and have their constraints satisfied.

Reply:
My mistake, I'd forgotten the meaning of "required" (as opposed to "implied") in the context of dimensions.